### PR TITLE
VACMS-12394: Prevent links with VA.gov domains as first path segment.

### LINKED ISSUE #12394: Prevent adding links with relative URLs having VA.gov domains as the first path segment.
## Description
E.g. prevent `<a href="/www.va.gov/some-page">links</a>` from being added to the system.

## Acceptance Criteria
- [ ] Validator exists for the above pattern.
- [ ] Unit tests exist for the validator.

### Team
Please check the team(s) that will do this work.

- [x] `CMS Team`
- [ ] `Public Websites`
- [ ] `Facilities`
- [ ] `User support`

### PR DESCRIPTION
**Pull request by ndouglas**
## Description

Closes #12394.

## QA steps

As any user with the ability to edit content:
1. Edit a node, e.g. [VAMC System Health Service Whole health at VA Cincinnati health care](https://cms-wpuulvbzdyrcqodsktt7i3ppei6scpm3.ci.cms.va.gov/node/24426/edit)
2. Insert a relative link containing a VA.gov domain into a rich text field.  For instance: `<a href="/www.va.gov/BUILD.txt">Example Link!</a>`.
3. Attempt to save the document.  You should see:
  - [x] An error message at the top of the screen:
Screen Shot 2023-02-01 at 11 02 56 AM

  - [ ] validate error message link jumps to the error.

  - [x] An error message at the field:
Screen Shot 2023-02-01 at 11 11 38 AM

  - [x] An error message at the bottom of the field:
Screen Shot 2023-02-01 at 11 11 43 AM

4. Edit a media image item, e.g. [August2022.jpg](https://cms-wpuulvbzdyrcqodsktt7i3ppei6scpm3.ci.cms.va.gov/media/image/16769)

5.  Insert a relative link containing a URL into a plain text field. For instance, `/www.va.gov/BUILD.txt`.

6. Attempt to save the media image item.  You should see:
  - [x] An error message at the top of the screen:
Screen Shot 2023-02-01 at 11 12 24 AM

  - [ ] validate error message link jumps to the error.

  - [x] An error message at the field:
Screen Shot 2023-02-01 at 11 12 28 AM

  - [x] An error message at the bottom of the field:
Screen Shot 2023-02-01 at 11 12 32 AM

### Definition of Done

- [ ] Documentation has been updated, if applicable.
- [ ] Tests have been added if necessary.
- [ ] Automated tests have passed.
- [ ] Code Quality Tests have passed.
- [ ] Acceptance Criteria in related issue are met.
- [ ] Manual Code Review Approved.

### CODE REVIEW

File: docroot/modules/custom/va_gov_backend/src/Plugin/Validation/Constraint/PreventVaGovDomainsAsPathsInLinksValidator.php [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+     *     [^\s/]*    Match any non-whitespace, non-slash characters...
+     *     va\.gov    ... and the exact string "va.gov"...
+     *     [^\s]*     ... and following non-whitespace characters.
+     *   )            Close capture group.
+     * #              End delimiter (replace '/')
+     *
+     * In other words, we look for a string like `/va.gov/something` or
+     * `/www.va.gov/something-else`.
+     */
+    if (preg_match('#[\s"\']/([^\s/]*va.gov[^\s]*)#', $text, $matches)) {
```
Comment (ndouglas) [inline docroot/modules/custom/va_gov_backend/src/Plugin/Validation/Constraint/PreventVaGovDomainsAsPathsInLinksValidator.php:84]:
```suggestion
    if (preg_match('#[\s"\']/([^\s/]*va\.gov[^\s]*)#', $text, $matches)) {
```

h/t @christianmburk

File: docroot/modules/custom/va_gov_backend/src/Plugin/Validation/Constraint/PreventVaGovDomainsAsPathsInLinksValidator.php [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+       *     [^\s/]*    Match any non-whitespace, non-slash characters...
+       *     va\.gov    ... and the exact string "va.gov"...
+       *     [^\s]*     ... and following non-whitespace characters.
+       *   )            Close capture group.
+       * #              End delimiter (replace '/')
+       *
+       * In other words, we look for a string like `/va.gov/something` or
+       * `/www.va.gov/something-else`, but not `/some-path/www.va.gov/`.
+       */
+      if (!preg_match('#^/([^\s/]*va.gov[^\s]*)#', $url)) {
```
Comment (ndouglas) [inline docroot/modules/custom/va_gov_backend/src/Plugin/Validation/Constraint/PreventVaGovDomainsAsPathsInLinksValidator.php:124]:
```suggestion
      if (!preg_match('#^/([^\s/]*va\.gov[^\s]*)#', $url)) {
```

**Review by swirtSJW — COMMENTED** [review]
Code looks great.  Love all the test cases.   I will run through manual QA when I am back at my desk.

**Comment by ndouglas** [comment]
@swirtSJW This has the same issue as the other validator, so I'm trying to figure out what changed.

**Comment by ndouglas** [comment]
Blocked by #12415.

**Comment by swirtSJW** [comment]
All tests out except the jump link. I vaguely recall running into this with one of Ethan's PRs but I don't recall the solution.

**Comment by swirtSJW** [comment]
@ndouglas since this is a core bug, I am ok with not blocking these validation issues on it.  We can't hold all this great work hostage for resolution of a core bug.

**Review by swirtSJW — APPROVED** [review]
Tested and approved.  I don't think we need to block these on the core defect.

**Pull request merged**